### PR TITLE
use bdimy = 1 to WAR smem race

### PR DESCRIPTION
**Pull request by liqiangxl**
when total_reduction_numel <= 1024, scheduler may use multiple reductions per block with bdimy > 1, this leads to race condition in shared memory when using async copy. Adding `cp.async.wait_all`after the 1st async copy can avoid the race, but needs to figure out the root cause before we can safely use it. So, here we put all buffers in registers.

race detected with:
```
NVFUSER_DUMP=scheduler_params,cuda_to_file NVFUSER_ENABLE=kernel_debug PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer --tool racecheck --racecheck-detect-level info  ./nvfuser_tests --gtest_filter='CombinedSchedulerTest.LayerNormBackward/dtype_double_batch_216_hidden_96'
```

### CODE REVIEW

**Comment by liqiangxl** [comment]
!test

**Comment by naoyam** [comment]
Is this WAR still a draft? I know you're working on a proper fix, but since it's a silent error, could you please prioritize landing this WAR first?

**Comment by liqiangxl** [comment]
> Is this WAR still a draft? I know you're working on a proper fix, but since it's a silent error, could you please prioritize landing this WAR first?

I already have a fix at https://github.com/NVIDIA/Fuser/pull/3438, if that looks reasonable, we don't need this WAR.

**Comment by naoyam** [comment]
It may take some time to review that PR, so let's get this merged for now.

File: csrc/scheduler/normalization_inner_outer.cpp [hunk 1]
```diff
@@ -279,6 +282,19 @@ PersistentBufferStorageParams getPersistentBufferStorageParams(
   buffer_params.smem_buffer_size = total_smem_buffer_size;
   buffer_params.regs_buffer_size =
       partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+
+  // when total_reduction_numel <= 1024, scheduler may use multiple reductions
+  // per block with bdimy > 1, this leads to race condition in shared memory
+  // when using async copy. Adding `cp.async.wait_all`after the 1st async copy
+  // can avoid the race, but needs to figure out the root cause before we can
+  // safely use it. So, here we put all buffers in registers.
+  if (total_reduction_numel <= 1024L) {
```
Comment (naoyam) [inline csrc/scheduler/normalization_inner_outer.cpp:291]: Does the issue happen when bdimy is greater than 1? If so, shouldn't we check the value of `bdimy` directly?

**Comment by liqiangxl** [comment]
!test

**Review by naoyam — APPROVED** [review]
LGTM

**Comment by liqiangxl** [comment]
`DistributedTransformerTest.MultiheadAttention_SP/__half`  fails at head

**Comment by liqiangxl** [comment]
!test

**Comment by liqiangxl** [comment]
!test

**Pull request merged**